src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,353 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Exporter.Geneva;
+
+internal class OtlpProtobufSerializer
+{
+    private const int LengthAndTagSize = 4;
+
+    private readonly Dictionary<string, List<Metric>> scopeMetrics = new();
+
+    private int resourceMetricStartIndex;
+
+    private int scopeMetricsStartIndex;
+
+    private int metricStartIndex;
+
+    private int intrumentStartIndex;
+
+    private int metricPointStartIndex;
+
+    internal IMetricDataTransport MetricDataTransport;
+
+    public OtlpProtobufSerializer(IMetricDataTransport metricDataTransport)
+    {
+        this.MetricDataTransport = metricDataTransport;
+    }
+
+    internal void SerializeAndSendMetrics(byte[] buffer, Resource resource, in Batch<Metric> metricBatch)
+    {
+        foreach (var metric in metricBatch)
+        {
+            if (this.scopeMetrics.TryGetValue(metric.MeterName, out var metricList))
+            {
+                metricList.Add(metric);
+            }
+            else
+            {
+                var newList = new List<Metric>();
+                newList.Add(metric);
+                this.scopeMetrics[metric.MeterName] = newList;
+            }
+        }
+
+        // Serialize
+        this.SerializeResourceMetrics(buffer, resource);
```

Comment:
Split this into SerializeResource and SerializeAllScopeMetrics. SerializeResource could be called only once and not in every iteration. You don’t need to free the entire buffer. Instead, provide the buffer and the length utilized earlier by resource to SerializeAllScopeMetrics

Reply:
I see, SerializeResourceMetrics already does the split and a TODO.